### PR TITLE
[Master]Support disk_pool to None by default

### PR DESCRIPTION
**Pull request by trenxiaof**
Please help review the code logic first, the ut and fvt will complete later on.
There's one point not quiet sure:
In guest_create() of api.py, if disk_list has only one disk info and disk_pool is None:
such as disk_list=[{'size':'1g', 'format':'xfs'}]  and disk_pool not configured in zvmsdk.conf
in such case, the guest_create report ERROR？or Continue( does boot_from_volume deploy vm go into this guest_create() fucntion in api.py?)

To support CIC manage a zvm host without diskpool, such as
zvm host configured FCP stroage only, we do the corresponding
change to support disk_pool to None by default.

Code changes:
- Add extra check for disk_pool=None before previous check logic
  > guest_create
  > host_diskpool_get_info
  > get_info
  > add_mdisks
  > _add_mdisk
- Not call Image_Volume_Space_Query_DM and return minimal disk values directly if disk_pool is None
- Add some comments for disk_pool is None
- Remove the disk from disk_list if disk_pool is None
- Add default value None to config option

Related issue:
https://github.ibm.com/zvc/planning/issues/2771

Signed-off-by: Xiao Feng Ren <renxiaof@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by trenxiaof** [comment]
@jichenjc @bjhuangr @dongyanyang @haoqf please help review.

File: zvmsdk/smtclient.py [hunk 1]
```diff
@@ -643,6 +651,10 @@ def _add_mdisk(self, userid, disk, vdev):
         size = disk['size']
         fmt = disk.get('format', 'ext4')
         disk_pool = disk.get('disk_pool') or CONF.zvm.disk_pool
+        # Check disk_pool, if it's None, do nothing
+        if disk_pool is None:
```
Comment (bjhuangr) [inline zvmsdk/smtclient.py:655]: duplicated check here

File: zvmsdk/smtclient.py [hunk 2]
```diff
@@ -166,11 +166,22 @@ def add_mdisks(self, userid, disk_list, start_vdev=None):
...
                        'format': 'ext3',
-                       'disk_pool': 'ECKD:eckdpool1'}]
+                       'disk_pool': 'ECKD:eckdpool1'},
+                       {'size': '1g',
+                       'format': 'ext3'}]
 
         """
 
         for idx, disk in enumerate(disk_list):
+            if disk.get('disk_pool') is None:
```
Comment (bjhuangr) [inline zvmsdk/smtclient.py:176]: if we have checked diskpool none in guest create and guest resize API, no need to check diskpool in smtclient, correct?

**Comment by trenxiaof** [comment]
@bjhuangr Thanks for the comments. I will update them.

**Review by SeanHQF — APPROVED** [review]

File: zvmsdk/smtclient.py (hunk 2 above)
Comment (trenxiaof) [reply zvmsdk/smtclient.py:176]: But it also checks CONF.zvm.disk_pool if disk_pool is None for add_mdisks() here. Maybe it still need check.

File: zvmsdk/api.py [hunk 3]
```diff
@@ -921,12 +942,18 @@ def guest_create_disks(self, userid, disk_list):
...
                In this case it will create one disk 0100(in case the vdev
                for root disk is 0100) with size 1g from ECKD disk pool
                eckdpool1 for guest , then set IPL 0100 in guest's user
                directory, and it will create 0101 with 200000 blocks from
-               FBA disk pool fbapool1, and formated with ext3.
+               FBA disk pool fbapool1, and formated with ext3. As for the third
+               case, if the disk_pool isn't configured in configure file, the
+               default value is None, the disk_pool here is None, no disk will
+               be created. If it's configured, such as ECKD:eckdpool2, it will
+               create 0102 with size 1g from ECKD disk pool eckdpool2 for guest.
```
[the remark this answers is not in the document]
Comment (trenxiaof) [reply zvmsdk/api.py:956]: not check here but in the sub-function, will change here.

File: zvmsdk/smtclient.py (hunk 1 above)
Comment (trenxiaof) [reply zvmsdk/smtclient.py:655]: but why it still check CONF.zvm.disk_pool here in original code?

**Comment by trenxiaof** [comment]
test with restclient:
>>> conn.send_request('guest_create_disks', 'WNGZHE22', [{'size': '1g', 'format': 'ext4'}])
{'overallRC': 100, 'modID': 400, 'rc': 100, 'rs': 3, 'errmsg': 'Invalid API arg format, error: Invalid disk_pool input, it should be configured for sdkserver.', 'output': ''}

>>> conn.send_request('host_diskpool_get_info')
{'overallRC': 0, 'modID': None, 'rc': 0, 'rs': 0, 'errmsg': '', 'output': {'disk_total': 0, 'disk_used': 0, 'disk_available': 0}}

>>> conn.send_request('host_get_info')
{'overallRC': 0, 'modID': None, 'rc': 0, 'rs': 0, 'errmsg': '', 'output': {'zcc_userid': 'WZBOE95', 'zvm_host': 'BOEIAAS1', 'vcpus': 12, 'vcpus_used': 12, 'cpu_info': {'architecture': 's390x', 'cec_model': '2964'}, 'memory_mb': 524288.0, 'memory_mb_used': 198348.8, 'hypervisor_type': 'zvm', 'hypervisor_version': 710, 'hypervisor_hostname': 'BOEIAAS1', 'ipl_time': 'IPL at 06/09/20 14:04:39 CET', 'disk_total': 0, 'disk_used': 0, 'disk_available': 0}}

>>> disks_list = [{'size': '2g', 'is_boot_disk': True}]
>>> conn.send_request('guest_create', userid, vcpus, memory, disk_list=disks_list)
{'overallRC': 100, 'modID': 400, 'rc': 100, 'rs': 3, 'errmsg': 'Invalid API arg format, error: Invalid disk_pool input, it should be configured for sdkserver.', 'output': ''}

**Comment by trenxiaof** [comment]
Use SDKGuestOperationError replace SDKFunctionNotImplementError, as the rs 2 of 'guest' errors is "2: "Failed to add mdisks when creating guest, error: %(msg)s"," it's suitable.

**Comment by bjhuangr** [comment]
please fix the pep8 error accordingly:
```
./zvmsdk/api.py:756:80: E501 line too long (80 > 79 characters)
./zvmsdk/api.py:797:80: E501 line too long (80 > 79 characters)
./zvmsdk/api.py:950:80: E501 line too long (80 > 79 characters)
./zvmsdk/tests/unit/test_api.py:114:80: E501 line too long (90 > 79 characters)
./zvmsdk/tests/unit/test_api.py:123:80: E501 line too long (87 > 79 characters)
./zvmsdk/tests/unit/test_api.py:253:80: E501 line too long (90 > 79 characters)
```

**Comment by trenxiaof** [comment]
> please fix the pep8 error accordingly:
> 
> ```
> ./zvmsdk/api.py:756:80: E501 line too long (80 > 79 characters)
> ./zvmsdk/api.py:797:80: E501 line too long (80 > 79 characters)
> ./zvmsdk/api.py:950:80: E501 line too long (80 > 79 characters)
> ./zvmsdk/tests/unit/test_api.py:114:80: E501 line too long (90 > 79 characters)
> ./zvmsdk/tests/unit/test_api.py:123:80: E501 line too long (87 > 79 characters)
> ./zvmsdk/tests/unit/test_api.py:253:80: E501 line too long (90 > 79 characters)
> ```

done

**Pull request merged**